logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java

```diff
@@ -279,14 +279,21 @@ public void validate() throws InvalidIRException {
             throw new InvalidIRException("Graph has no leaf vertices!\n" + this.toString());
         }
 
-        List<List<Vertex>> duplicates = verticesByHash().values().stream().filter((group) -> group.size() > 1).collect(Collectors.toList());
-        if (!duplicates.isEmpty()) {
-            Stream<String> errorMessageGroups = duplicates.stream().
-                    map((group) -> group.stream().map(Object::toString).collect(Collectors.joining("===")));
-
-            String joinedErrorMessageGroups = errorMessageGroups.collect(Collectors.joining("\n---\n"));
-
-            throw new InvalidIRException("Some nodes on the graph are fully redundant!\n" + this + "|" + joinedErrorMessageGroups);
+        // Check for duplicate IDs in the config
+        List<String> duplicateIdErrorMessages = this.vertices().parallel()
```

Comment:
Yes, `getId()` can be expensive since it usually has to compute the hash which is expensive. The parallel does work as indicated above unless I'm mistaken and my previous benchmark had a mistake in it.